Dockerfile

```diff
@@ -1,31 +1,55 @@
-FROM rustlang/rust:nightly AS builder
-RUN apt update && apt install -y npm && rm -rf /var/apt/lists
-ARG GITHUB_SHA
-ENV GITHUB_SHA=$GITHUB_SHA
-ARG GITHUB_REPOSITORY
-ENV GITHUB_REPOSITORY=$GITHUB_REPOSITORY
+# Setup chef
+FROM rustlang/rust:nightly AS chef
+RUN cargo install cargo-chef
+
+# Build
+FROM chef AS planner
 WORKDIR /app
-COPY Cargo.toml Cargo.lock package.json ./
-RUN npm i
-RUN mkdir src && echo "fn main(){println!(\"dummy\");}" > ./src/main.rs && touch src/lib.rs
-RUN mkdir src/bin && echo "fn main(){println!(\"dummy\");}" > ./src/bin/bell_mkschema.rs
-RUN cargo build --release --bin ethsbell-rewrite --features=ws
-RUN rm ./src/main.rs && rm ./src/lib.rs
-COPY . .
-RUN touch src/main.rs && touch src/lib.rs
-# This is here so the build fails if the tests do.
+COPY Cargo.toml Cargo.lock ./
+RUN cargo chef prepare --recipe-path recipe.json
+
+FROM chef AS builder
+WORKDIR /app
+COPY --from=planner /app/recipe.json recipe.json
+RUN cargo chef cook --release --recipe-path recipe.json
+
+# Build image and run tests
+COPY LICENSE.html def.json .
+COPY src src
+COPY tests tests
+COPY templates templates
 RUN cargo test --release --features=ws
-RUN npm run build
 RUN cargo build --release --bin ethsbell-rewrite --features=ws
 
+FROM ubuntu AS frontend
+RUN apt-get update && apt-get install -y libssl-dev ca-certificates curl make build-essential && rm -rf /var/lib/apt/lists
+RUN curl -fsSL https://deb.nodesource.com/setup_14.x | bash -
+RUN apt-get install -y nodejs
+WORKDIR /app
+COPY package*.json .posthtmlrc .
+RUN npm i
+COPY frontend frontend
+RUN npm run build -- --no-cache
+COPY frontend/favicon.ico frontend-dist
+
 FROM ubuntu
 WORKDIR /app
-RUN apt-get update && apt-get install -y libssl-dev ca-certificates curl && rm -rf /var/lib/apt/lists
+
+RUN apt-get update && apt-get install -y libssl-dev ca-certificates && rm -rf /var/lib/apt/lists
 RUN update-ca-certificates
-COPY --from=builder /app/def.json /app/target/release/ethsbell-rewrite ./
-COPY --from=builder /app/def.d/* ./def.d/
-COPY --from=builder /app/frontend-dist ./frontend-dist
-COPY --from=builder /app/frontend/favicon.ico /app/frontend-dist/favicon.ico
-COPY --from=builder /app/templates ./templates
+RUN apt update && rm -rf /var/apt/lists
+ARG GITHUB_SHA
+ENV GITHUB_SHA=$GITHUB_SHA
+ARG GITHUB_REPOSITORY
+ENV GITHUB_REPOSITORY=$GITHUB_REPOSITORY
+
+# Rust
+COPY --from=builder /app/target/release/ethsbell-rewrite .
+COPY def.d def.d
```

Comment:
This will fail because you removed the default schedule files.

Reply:
Fixed